crates/kitsune_p2p/kitsune_p2p/tests/sharding.rs

```diff
@@ -0,0 +1,424 @@
+use crate::common::{
+    start_bootstrap, start_signal_srv, wait_for_connected, KitsuneTestHarness, TestHostOp,
+};
+use fixt::fixt;
+use kitsune_p2p::actor::{BroadcastData, KitsuneP2pSender};
+use kitsune_p2p::dht::arq::LocalStorageConfig;
+use kitsune_p2p::dht::prelude::SpaceDimension;
+use kitsune_p2p::dht_arc::DhtLocation;
+use kitsune_p2p_bin_data::fixt::KitsuneSpaceFixturator;
+use kitsune_p2p_bin_data::{KitsuneAgent, KitsuneBasis, KitsuneBinType};
+use kitsune_p2p_fetch::FetchContext;
+use kitsune_p2p_types::config::tuning_params_struct;
+use kitsune_p2p_types::dht::{Arq, ArqStrat};
+use kitsune_p2p_types::KitsuneTimeout;
+use num_traits::AsPrimitive;
+use std::sync::Arc;
+
+mod common;
+
+/// Test scenario steps:
+///   1. Set up 5 nodes, each with one agent.
+///   2. Assign a DHT arc to each agent such that they overlap with the next agent's start location.
+///   3. Connect each agent to the next agent (circular), so that we know they are aware of each other.
+///   4. Publish an op with a basis location set to the location of the 4th agent. This should also be visible to the 3rd agent by 2. above.
+///   5. Wait for the 3rd to receive the data.
+///   6. Assert that the op was never published to the 1st, 2nd, or 5th agents. (Note that we cannot check if we sent it to ourselves because the op was already in our store)
+#[cfg(feature = "tx5")]
+#[tokio::test(flavor = "multi_thread")]
+async fn publish_to_basis_from_inside() {
+    holochain_trace::test_run();
+
+    let (bootstrap_addr, _bootstrap_handle) = start_bootstrap().await;
+    let (signal_url, _signal_srv_handle) = start_signal_srv().await;
+
+    let space = Arc::new(fixt!(KitsuneSpace));
+
+    let tuner = |mut params: tuning_params_struct::KitsuneP2pTuningParams| {
+        params.gossip_arc_clamping = "none".to_string();
+        params.gossip_dynamic_arcs = false; // Don't update the arcs dynamically, use the initial value
+        params.disable_recent_gossip = true;
+        params.disable_historical_gossip = true;
+        params
+    };
+
+    // Arcs are this long by default, with an adjustment to ensure overlap.
+    let base_len = u32::MAX / 5;
+
+    let dim = SpaceDimension::standard();
+
+    let mut agents = Vec::new();
+
+    for i in 0..5 {
+        let mut harness = KitsuneTestHarness::try_new("")
```

Comment:
@neonphog I forget, are we intending to keep using KitsuneTestHarness or should we use something else?

Reply:
This is the new test harness that @ThetaSinner wrote, not to be confused with the old test harness that I wrote : )

Reply:
Yep, fairly recently added still and although it is enabling the writing of some useful tests and is a decent match to the 'real' host behaviour, there seem to be issues with flakiness on CI that I don't see locally. It's unclear to me yet whether that's a problem with the harness or is some difference in behaviour of Kitsune when running with fewer resources